corporate/lib/stripe.py

```diff
@@ -4437,7 +4442,9 @@ def get_plan_renewal_or_end_date(plan: CustomerPlan, event_time: datetime) -> da
 def invoice_plans_as_needed(event_time: Optional[datetime] = None) -> None:
     if event_time is None:  # nocoverage
         event_time = timezone_now()
-    for plan in CustomerPlan.objects.filter(next_invoice_date__lte=event_time):
+    for plan in CustomerPlan.objects.filter(
+        Q(next_invoice_date__lte=event_time) | Q(end_date__lte=event_time)
+    ):
```

Comment:
Does this need to filter on the status being live? I don't think we want to have this code be called for plans whose status is `ENDED` and have an end date in the past.

Reply:
I guess also we really only need to call `self.make_end_of_cycle_updates_if_needed`, and not invoicing, if there's no `next_invoice_date`; I wonder if it'd be better to have that explicitly happen -- maybe just a block to do that and then `continue` -- we don't really want to be creating `InvoiceItem` objects or anything for legacy plan customers and I don't really want to have to audit `invoice_plan` for being safe to call with a plan that is not billable.

so something like:

```
if plan.next_invoice_date is None:
     # Handle plan updates for plans who end date is in the past.
     self.make_end_of_cycle_updates_if_needed(plan, event_time)
     continue
```

Reply:
> I guess also we really only need to call `self.make_end_of_cycle_updates_if_needed`, and not invoicing

We were not invoicing as our `invoiced_through_id` is already up-to-date with the last ledger entry. But using a `continue` will save us a DB query  -- so used a `continue` block.

Reply:
> Does this need to filter on the status being live? I don't think we want to have this code be called for plans whose status is ENDED and have an end date in the past.

We also need to invoice plans having status = `NEVER_STARTED`, so the first impression would be to use `~Q(status=ENDED)`.

But the problem is, In `test_downgrade`:
```python
        # Verify that we call invoice_plan once more after cycle end but
        # don't invoice them for users added after the cycle end
        plan = CustomerPlan.objects.first()
        assert plan is not None
        self.assertIsNotNone(plan.next_invoice_date)
```
So, we invoice a Plan with `status=ENDED` & not None `next_invoice_date` for once. And then the next_invoice_date is set to None.

This is the same case we discussed in https://github.com/zulip/zulip/pull/28660#discussion_r1462223863

Should we just stick to involve `next_invoice_date` as the deciding factor for "which plans to invoice" instead of involving end_date & status?

Just invoice legacy plan for once (to downgrade via make_end_of_cycle...) & set `next_scheduled_date` to None. -- We can add tests to verify "functions to create invoice are not called".

Reply:
Yeah I think such tests would probably be sufficient for safety here.